commit 09e1ad759b21
Author: Hiroyuki Ikezoe <hikezoe@mozilla.com>
Date:   Wed Oct 18 10:43:22 2017 +0900

    Bug 1385013 - Check all vertexes for the target frame are outside of the parent frame if the target frame is empty. r=birtles, a=ritu
    
    We create empty rectangle (zero-height or zero-width) frame for element which
    has no content inside it, e.g.  '<p></p>'.  And nsRect.Intersects returns false
    if either of the rectangles are empty, so if we check
    !transformedRect.Intersects(scrollableRect) and transformedRect is empty, we
    will end up returning true from IsFrameScrolledOutOfView even though the point
    represented by the empty transformedRect might be inside the
    scrollableRect.
    
    The reftest causes timeout without this fix since the animation never updates
    after the initial paint.
    
    MozReview-Commit-ID: FymFJfjxyGc
    
    --HG--
    extra : source : 69e2ab081d3d71ef131f0ce878ee7830ae881d08
---
 layout/generic/nsFrame.cpp                         | 11 +++++++++-
 .../animation-on-empty-height-frame.html           | 25 ++++++++++++++++++++++
 layout/reftests/css-animations/reftest.list        |  1 +
 3 files changed, 36 insertions(+), 1 deletion(-)

diff --git layout/generic/nsFrame.cpp layout/generic/nsFrame.cpp
index dbfd61b7e142..991ccbceeb65 100644
--- layout/generic/nsFrame.cpp
+++ layout/generic/nsFrame.cpp
@@ -10555,7 +10555,16 @@ IsFrameScrolledOutOfView(nsIFrame *aFrame)
                                                 scrollableParent);
 
   nsRect scrollableRect = scrollableParent->GetVisualOverflowRect();
-  if (!transformedRect.Intersects(scrollableRect)) {
+  if (transformedRect.IsEmpty()) {
+    // If the transformed rect is empty it represents a line or a point that we
+    // should check is outside the the scrollable rect.
+    if (transformedRect.x > scrollableRect.XMost() ||
+        transformedRect.y > scrollableRect.YMost() ||
+        scrollableRect.x > transformedRect.XMost() ||
+        scrollableRect.y > transformedRect.YMost()) {
+      return true;
+    }
+  } else if (!transformedRect.Intersects(scrollableRect)) {
     return true;
   }
 
diff --git layout/reftests/css-animations/animation-on-empty-height-frame.html layout/reftests/css-animations/animation-on-empty-height-frame.html
new file mode 100644
index 000000000000..eb3f58127b16
--- /dev/null
+++ layout/reftests/css-animations/animation-on-empty-height-frame.html
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
+<html class="reftest-wait reftest-no-flush">
+<style>
+@keyframes anim {
+  from { background-color: white; }
+  to { background-color: red; }
+}
+</style>
+<body>
+</body>
+<script>
+window.addEventListener('load', () => {
+  const body = document.querySelector('body');
+  body.style.animation = 'anim 100s step-end reverse';
+  body.addEventListener('animationstart', () => {
+    // This MozAfterPaint event corresponds to the white background paint.
+    // (The animation will initially paint the background red since it is playing
+    // a step-end animation in reverse.)
+    window.addEventListener('MozAfterPaint', () => {
+      document.documentElement.classList.remove('reftest-wait');
+    }, {once: true});
+  });
+});
+</script>
+</html>
diff --git layout/reftests/css-animations/reftest.list layout/reftests/css-animations/reftest.list
index 61e3e9ea6101..ed615645f707 100644
--- layout/reftests/css-animations/reftest.list
+++ layout/reftests/css-animations/reftest.list
@@ -4,6 +4,7 @@ skip-if(stylo) fails == print-no-animations.html print-no-animations-ref.html #
 skip-if(stylo) fails != print-no-animations.html print-no-animations-notref.html # reftest harness doesn't actually make pres context non-dynamic for reftest-paged tests. Bug 1374154 for stylo
 == animate-opacity.html animate-opacity-ref.html
 == animate-preserves3d.html animate-preserves3d-ref.html
+== animation-on-empty-height-frame.html about:blank
 == in-visibility-hidden-animation.html in-visibility-hidden-animation-ref.html
 == in-visibility-hidden-animation-pseudo-element.html in-visibility-hidden-animation-pseudo-element-ref.html
 == partially-out-of-view-animation.html partially-out-of-view-animation-ref.html
